service/service.go

```diff
@@ -170,16 +169,75 @@ func NewService(sig *signal.Signal, cfg *Config, datadir string) (*Service, erro
 	return s, nil
 }
 
+func (s *Service) loop(ctx context.Context) {
+	executingRounds, _ := errgroup.WithContext(ctx)
+	defer executingRounds.Wait()
+
+	timer := time.NewTimer(0)
+	<-timer.C
+	defer timer.Stop()
+	for {
+		start := s.genesis.
+			Add(s.cfg.EpochDuration * time.Duration(s.openRound.Epoch())).
+			Add(s.cfg.PhaseShift)
+		if d := time.Until(start); d > 0 {
+			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), d)
+			timer.Reset(d)
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				log.Info("service shutting down")
+				s.openRoundMutex.Lock()
+				s.openRound = nil
+				s.openRoundMutex.Unlock()
+				return
+			}
+		}
+
+		s.openRoundMutex.Lock()
+		if s.openRound.isEmpty() && !s.cfg.ExecuteEmpty {
+			s.openRoundMutex.Unlock()
+			continue
+		}
+
+		prevRound := s.openRound
+		open := time.Now()
+		s.newRound(ctx, prevRound.Epoch()+1)
+		s.openRoundMutex.Unlock()
+		log.Info("Round %v opened. took %v", s.openRoundID(), time.Since(open))
+
+		executingRounds.Go(func() error {
+			round := prevRound
+			if err := s.executeRound(ctx, round); err != nil {
+				s.asyncError(fmt.Errorf("round %v execution error: %v", round.ID, err))
+				return nil
+			}
+			broadcastProof(s, round, round.execution, s.getBroadcaster())
+			return nil
+		})
+	}
+}
+
 func (s *Service) Start(b Broadcaster) error {
-	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
+	s.Lock()
+	if !s.started.CompareAndSwap(false, true) {
+		s.Unlock()
 		return ErrAlreadyStarted
 	}
 
+	// Create a context for the running Service.
+	// This context will be canceled when Service is stopped.
+	ctx, stop := context.WithCancel(context.Background())
+	runningGroup, _ := errgroup.WithContext(ctx)
```

Comment:
Hm... I'm not sure about this. The idea of the `context` returned by `errgroup.WithContext` is that it will be canceled by the `errgroup` automatically if one of the go routines insides returns an error. So the context returned by the `errgroup` imho should be passed into `runningGroup.Go(func() {})` over any other context, since otherwise the other routines don't stop as early as possible.

We can still use the parent context to cancel explicitly, but the routines started with `runningGroup.Go` should listen to the context returned by `errgroup.WithContext`.

Reply:
As discussed offline - I specifically don't want the auto-cancellation feature. As suggested, I refactored to use a zero-value error group without context.

The spawned goroutines check the parent context (which describes the service running lifetime) for cancellation.